packages/gatsby-plugin-sharp/src/index.js

```diff
@@ -95,60 +126,181 @@ function queueImageResizing({ file, args = {}, reporter }) {
   const prefixedSrc =
     options.pathPrefix + `/static/${digestDirPrefix}` + encodedImgSrc
 
-  // Create job and add it to the queue, the queue will be processed inside gatsby-node.js
-  const job = {
-    args: options,
-    inputPath: file.absolutePath,
-    contentDigest: file.internal.contentDigest,
-    outputDir,
-    outputPath: outputFilePath,
+  return {
+    src: prefixedSrc,
+    outputDir: outputDir,
+    relativePath: outputFilePath,
+    width,
+    height,
+    aspectRatio,
+    options,
   }
+}
 
-  const outputFile = path.join(job.outputDir, job.outputPath)
-  let finishedPromise = Promise.resolve()
+async function createJob(job, { reporter }) {
+  const progressBar = createOrGetProgressBar(reporter)
 
-  if (cachedOutputFiles.has(outputFile)) {
-    finishedPromise = cachedOutputFiles.get(outputFile)
+  if (pendingImagesCounter === 0) {
+    progressBar.start()
   }
 
-  // Check if the output file already exists or already is being created.
-  // TODO: Remove this when jobs api is stable, it will have a better check
-  if (!fs.existsSync(outputFile) && !cachedOutputFiles.has(outputFile)) {
-    // schedule job immediately - this will be changed when image processing on demand is implemented
-    finishedPromise = scheduleJob(
-      job,
-      boundActionCreators,
-      pluginOptions,
-      reporter
-    ).then(res => {
-      cachedOutputFiles.delete(outputFile)
+  const transforms = job.args.operations
+  pendingImagesCounter += transforms.length
+  progressBar.total = pendingImagesCounter
 
-      return res
-    })
-
-    cachedOutputFiles.set(outputFile, finishedPromise)
+  if (boundActionCreators.createJobV2) {
+    await boundActionCreators.createJobV2(job)
+  } else {
+    await scheduleJob(job, boundActionCreators)
   }
 
+  progressBar.tick(transforms.length)
+}
+
+function queueImageResizing({ file, args = {}, reporter }) {
+  const {
+    src,
+    width,
+    height,
+    aspectRatio,
+    relativePath,
+    outputDir,
+    options,
+  } = prepareQueue({ file, args })
+
+  // Create job and add it to the queue, the queue will be processed inside gatsby-node.js
+  const finishedPromise = createJob(
+    {
+      name: IMAGE_PROCESSING_JOB_NAME,
+      inputPaths: [file.absolutePath],
+      outputDir,
+      args: {
+        operations: [
+          {
+            outputPath: relativePath,
+            args: [options],
```

Comment:
```suggestion
            args: options,
```

(related to `resize` function not doing actual resize because `args` is array and not an object):
![Screenshot 2020-01-13 at 16 54 11](https://user-images.githubusercontent.com/419821/72279546-a570fa00-3636-11ea-8082-56a178ffea22.png)

Reply:
Not sure if good idea, but maybe `process-file` (or somewhere else in the chain) we could validate the shape of arguments? TS could help as here, but I don't think we will convert in this PR :)

Reply:
I could do joi validation but technically this should just be tested by a unit test. I feel like the plugin-sharp should just have a better suite. I can make a ticket and probably work on it later. If we test index.js -> processfile with only mocking sharp that would fix a lot.

Reply:
Up to you!